src/com/spike/RecursiveDescentParser.java

```diff
@@ -1,16 +1,397 @@
+    }
+
+    throw new AssertionError("Unreachable");
+  }
+
+  private Expr call() {
+    Expr value = primary();
+
+    if (match(LEFT_PAREN)) {
+      List<Expr> args = new ArrayList<>();
+
+      do {
+        args.add(expr());
+      } while (match(COMMA));
+
+      Token paren = consume(RIGHT_PAREN, "')' is missing after call expression");
+      return new Expr.Call(value, paren, args);
+    }
+
+    return value;
+  }
+
+  private Expr unary() {
+    if (!is(NOT, MINUS, BIT_COMPL)) {
+      return call();
+    }
+
+    return new Expr.Unary(advance(), unary());
+  }
+
+  // TODO: reduce boilerplate
+  private Expr multiplication() {
+    Expr left = unary();
+
+    while (is(SLASH, STAR, REMAINDER)) {
+      left = new Expr.Binary(left, advance(), unary());
+    }
+
+    return left;
+  }
+
+  private Expr addition() {
+    Expr left = multiplication();
+
+    while (is(PLUS, MINUS)) {
+      left = new Expr.Binary(left, advance(), multiplication());
+    }
+
+    return left;
+  }
+
+  private Expr shift() {
+    Expr left = addition();
+
+    while (is(BIT_LEFT, BIT_RIGHT)) {
+      left = new Expr.Binary(left, advance(), addition());
+    }
+
+    return left;
+  }
+
+  private Expr comparison() {
+    Expr left = shift();
+
+    while (is(LESS, LESS_EQUAL, GREATER, GREATER_EQUAL)) {
+      left = new Expr.Binary(left, advance(), shift());
+    }
+
+    return left;
+  }
+
+  private Expr equality() {
+    Expr left = comparison();
+
+    while (is(EQUAL_EQUAL, NOT_EQUAL)) {
+      left = new Expr.Binary(left, advance(), comparison());
+    }
+
+    return left;
+  }
+
+  private Expr bitwiseAnd() {
+    Expr left = equality();
+
+    while (is(BIT_AND)) {
+      left = new Expr.Binary(left, advance(), equality());
+    }
+
+    return left;
+  }
+
+  private Expr bitwiseXor() {
+    Expr left = bitwiseAnd();
+
+    while (is(BIT_XOR)) {
+      left = new Expr.Binary(left, advance(), bitwiseAnd());
+    }
+
+    return left;
+  }
+
+  private Expr bitwiseOr() {
+    Expr left = bitwiseXor();
+
+    while (is(BIT_OR)) {
+      left = new Expr.Binary(left, advance(), bitwiseXor());
+    }
+
+    return left;
+  }
+
+  private Expr logicalAnd() {
+    Expr left = bitwiseOr();
+
+    while (is(AND)) {
+      left = new Expr.Binary(left, advance(), bitwiseOr());
+    }
+
+    return left;
+  }
+
+  private Expr logicalOr() {
+    Expr left = logicalAnd();
+
+    while (is(OR)) {
+      left = new Expr.Binary(left, advance(), logicalAnd());
+    }
+
+    return left;
+  }
+
+  // FIXME later when classes will be introduced
+  private Expr assignment() {
+    if (!isNext(
+        EQUAL, PLUS_EQUAL, MINUS_EQUAL, STAR_EQUAL, SLASH_EQUAL, REMAINDER_EQUAL)) {
+      return logicalOr();
+    }
+
+    Token name = consume(IDENTIFIER, "Assignment target is missing");
+    Token op = advance();
+    Expr value = assignment();
+
+    if (op.type != EQUAL) {
+      value = new Expr.Binary(new Expr.Variable(name), op, value);
```

Comment:
Here I think the `op` should store default operator (`+`, `-`, `*` etc.) instead of `+=`, `-=`, etc. So the expression:
`x += 5` will be deconstructed to `Assign(x, Binary(x, '+', 5))`. Therefore it will allow to skip this step at Interpreter's part and make it more clear.

Reply:
The issue here is that we are forced to create dummy `+`, `-` etc. operators. It doesn't look good to me. And it might affect error reporting in the future (since those dummy operators will have invalid column and line number) though that's not so likely.